crates/cxx-qt-lib/src/core/qtlogging.rs

```diff
@@ -0,0 +1,139 @@
+// SPDX-FileCopyrightText: 2024 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Joshua Goins <joshua.goins@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+use cxx::{type_id, ExternType};
+use std::ffi::c_char;
+use std::ffi::CStr;
+use std::marker::PhantomData;
+
+#[cxx::bridge]
+mod ffi {
+    #[repr(i32)]
+    enum QtMsgType {
+        QtDebugMsg = 0,
+        QtInfoMsg = 4,
+        QtWarningMsg = 1,
+        QtFatalMsg = 3,
+        QtCriticalMsg = 2,
+        QtSystemMsg = 2
+    }
+
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/common.h");
+
+        include!("cxx-qt-lib/qstring.h");
+        type QString = crate::QString;
+
+        include!("cxx-qt-lib/qtlogging.h");
+        type QMessageLogContext<'a> = crate::QMessageLogContext<'a>;
+        type QtMsgType;
+
+        fn qt_message_output(msgType: QtMsgType, context: &QMessageLogContext, string: &QString);
+
+        #[cxx_name = "qmessagelogcontext_line"]
+        fn line(context: &QMessageLogContext) -> i32;
+
+        #[cxx_name = "qmessagelogcontext_set_line"]
+        fn set_line(context: &mut QMessageLogContext, line: i32);
+
+        #[cxx_name = "qmessagelogcontext_file"]
+        unsafe fn file(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_file"]
+        unsafe fn set_file(context: &mut QMessageLogContext, file: *const c_char);
+
+        #[cxx_name = "qmessagelogcontext_function"]
+        unsafe fn function(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_function"]
+        unsafe fn set_function(context: &mut QMessageLogContext, function: *const c_char);
+
+        #[cxx_name = "qmessagelogcontext_category"]
+        unsafe fn category(context: &QMessageLogContext) -> *const c_char;
+
+        #[cxx_name = "qmessagelogcontext_set_category"]
+        unsafe fn set_category(context: &mut QMessageLogContext, category: *const c_char);
+    }
+
+    #[namespace = "rust::cxxqtlib1"]
+    unsafe extern "C++" {
+        #[doc(hidden)]
+        #[rust_name = "qmessagelogcontext_default"]
+        fn construct() -> QMessageLogContext<'static>;
+    }
+}
+
+#[repr(C)]
+pub struct QMessageLogContext<'a> {
+    version: i32,
+    line: i32,
+    file: &'a *const c_char,
+    function: &'a *const c_char,
```

Comment:
This is not the correct type!

`&*const c_char` is a reference to a pointer, so similar to a `const c_char**`.
So reading this from Rust after being assigned in C++ is undefined behavior, as C++ thinks it's a `const char*`.

This needs to just be a `*const c_char`.

To introduce the required lifetime, you can add a single `PhantomData<&'a c_char>` member to the struct.
I'm a bit uncertain how that interacts with `#[repr(C)]`. In `#[repr(Rust)]` the `PhantomData` won't change the size of the struct, but in `#[repr(C)]` it might add a single byte, which unfortunately wouldn't allow us to make this type trivial...
Please use static assertions to check what happens either way :)

Reply:
Added, it seems like PhantomData does not change the memory layout of the struct - it's still 32 bytes. However, how do I add a static_assert for this? It looks like all of the ones in the .cpp files are checking the size of the Qt type, not the Rust type?